### PR TITLE
Ingest Tampa (THEA) Connected Vehicle (CV) Pilot data.

### PR DESCRIPTION
**Pull request by chuehlien**
The following changes were made to support the ingestion of data from Tampa (THEA) CV Pilot (https://www.tampacvpilot.com/) into ITS DataHub's Sandbox S3 bucket and into corresponding Socrata data sets on data.transportation.gov (DTG).

The ingestion involves two steps: when THEA deposit the data files to DataHub's ingestion s3 bucket, a lambda function will unzip the gzip file deposited by THEA and write the data to appropriate folders in the Sandbox S3 bucket based on data type and data generated time. Once in the Sandbox S3 bucket, a separate lambda function will transform the data appropriately according to its data type and write the data to the appropriate data sets on DTG. 

Data transformation includes flattening of the data structure, which is
required for the data to work with the Socrata backend on data.transportation.gov.
Renaming of certain fields is done to achieve consistency across data sets. No
unit conversion will be done to the data, though occasionally, additional fields
will be added to enhance usage of the dataset in Socrata (e.g. geolocation fields
for plotting and random number fields to allow random sampling). Any added fields will be listed in the description of each Socrata data set.

The data that we received from THEA include Basic Safety Message (BSM), Traveler Information Message (TIM), and Signal Phasing and Timing (SPaT) data. Examples of these can be found in the `/sandbox/samples/` directory of this repo.

Corresponding Socrata data sets include:
BSM: https://data.transportation.gov/Automobiles/Tampa-CV-Pilot-Basic-Safety-Message-BSM-Sample/nm7w-nvbm
TIM: https://data.transportation.gov/Automobiles/Tampa-CV-Pilot-Traveler-Information-Message-TIM-Sa/in46-gmir
SPaT: https://data.transportation.gov/Automobiles/Tampa-CV-Pilot-Signal-Phasing-and-Timing-SPaT-Samp/xn7c-yu2n

Note that the data sets will not be publicly available until March 5th.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,21 +123,25 @@ To limit the data being dowloaded you can use AWS CLI's filtering which is detai
...
 
-
-
 ### Wyoming CV Data
 
+Wyoming (WYDOT) currently provides sanitized BSM and TIM data to the public through ITS DataHub.
+
+- The full sanitized WYDOT BSM and TIM data set can be found in the [ITS DataHub sandbox s3 bucket](http://usdot-its-cvpilot-public-data.s3.amazonaws.com/index.html) under `wydot/` folder prefix
+- Sample WYDOT CV data sets on data.transportation.gov can be found [here](https://data.transportation.gov/browse?tags=wyoming%20connected%20vehicle%20%28cv%29%20pilot)
 - [Details on Wyoming CV DATA BSMs and TIMs messages and samples](https://github.com/usdot-jpo-ode/jpo-ode/blob/master/docs/ODE_Output_Schema_Reference.docx)
```
Comment (hmusavi) [inline README.md]: The link and the description of the link don't match. The link takes you to ODE reference guide.

File: README.md [hunk 2]
```diff
@@ -126,21 +123,25 @@ To limit the data being dowloaded you can use AWS CLI's filtering which is detai
...
-- [Full data set in AWS](http://usdot-its-cvpilot-public-data.s3.amazonaws.com/index.html)
 
-#### WYDOT BSM
+### Tampa CV Data
 
-- [Sample Data](https://data.transportation.gov/Automobiles/Wyoming-CV-Pilot-Basic-Safety-Message-One-Day-Samp/9k4m-a3jc)
+Tampa (THEA) currently provides sanitized BSM, TIM, and SPaT data to the public through ITS DataHub
+
+- The full sanitized THEA BSM, TIM, and SPaT data set can be found in the [ITS DataHub sandbox s3 bucket](http://usdot-its-cvpilot-public-data.s3.amazonaws.com/index.html) under `thea/` folder prefix
+- Sample THEA CV data sets on data.transportation.gov can be found [here](https://data.transportation.gov/browse?tags=tampa%20connected%20vehicle%20%28cv%29%20pilot)
```
Comment (hmusavi) [inline README.md]: The link doesn't pull any results, even when logged in which it should even when the dataset is private, right?

File: source/lambda_functionbsm_thea.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        ('payload_data_coreData_', 'coreData_'),
+        ('coreData_accelSet_', 'coreData_accelset_')
+    ]
+    rename_fields = [
+        ('metadata_dataType', 'dataType'),
+        ('metadata_recordGeneratedAt', 'metadata_generatedAt'),
+        ('metadata_recordGeneratedBy', 'metadata_generatedBy'),
+        ('coreData_lat', 'coreData_position_lat'),
+        ('coreData_long', 'coreData_position_long'),
+        ('coreData_elev', 'coreData_elevation'),
```
Comment (hmusavi) [inline source/lambda_functionbsm_thea.py:81]: Shouldn't elevation also be coreData_position_elevation for consistency?

File: source/lambda_functionbsm_thea.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        ('coreData_accelset_yaw', 'coreData_accelset_accelYaw')
```
Comment (hmusavi) [inline source/lambda_functionbsm_thea.py:82]: What about the other accelset elements: pitch, and roll?

**Review by hmusavi — APPROVED** [review]
Approved but with requested changes since these functions are currently running in production. Also, I did not review the logic of the functions very deeply. A deeper and more thorough review of the functions is advised. Also, unit tests should be implemented to test and validate the logic.

File: source/lambda_functionbsm_thea.py (hunk 3 above)
Comment (chuehlien) [reply source/lambda_functionbsm_thea.py:81]: `coreData_elev` was renamed to `coreData_elevation` to be consistent with the field name in WYDOT BSM (https://data.transportation.gov/d/9k4m-a3jc).

File: source/lambda_functionbsm_thea.py (hunk 4 above)
Comment (chuehlien) [reply source/lambda_functionbsm_thea.py:82]: The accelset element in BSM core uses "AccelerationSet4Way" and only contains long, lat, vert, and yaw, according to the J2735_201603.pdf document. Let me know if I should be referencing another schema standard document.

File: README.md (hunk 2 above)
Comment (chuehlien) [reply README.md]: I've updated the link in the README to point to the correct Tampa CVP tag: https://data.transportation.gov/browse?tags=tampa+connected+vehicle+pilot+deployment+%28tampa+cv+pilot%29&utf8=%E2%9C%93

File: README.md (hunk 1 above)
Comment (chuehlien) [reply README.md]: This has bee updated to the following:

- Wyoming CV BSM/TIM schema descriptions can be found in the [ODE Output Schema Reference](https://github.com/usdot-jpo-ode/jpo-ode/blob/master/docs/ODE_Output_Schema_Reference.docx) document

File: source/lambda_functionbsm_thea.py (hunk 3 above)
Comment (hmusavi) [reply source/lambda_functionbsm_thea.py:81]: Yeah, that's consistent with Wyoming but Wyoming's naming is inconsistent with the other `position` elements. Latitude, longitude and elevation are three components of a 3D position so they should all have the word `position` in their name but I understand your point. We'll leave it as is to be consistent with Wyoming.

File: source/lambda_functionbsm_thea.py (hunk 4 above)
Comment (hmusavi) [reply source/lambda_functionbsm_thea.py:82]: Correct, but `coreData_position_lat`, `coreData_position_long` and `coreData_elevation' are not part of `coreData_accelset`. I believe we need total of 7 elements:

1. coreData_position_lat
2. coreData_position_lon
3. coreData_position_elev
4. coreData_accelset_lat
5. coreData_accelset_lon
6. coreData_accelset_vert
7. coreData_accelset_yaw

Comment (chuehlien) [reply source/lambda_functionbsm_thea.py:82]: Adding notes from our side discussion:
We do have the corresponding elements for the 7 fields listed above in the Tampa BSM dataset. 

1. coreData_position_lat
2. coreData_position_long
3. coreData_elevation (named this way to be consistent with WYDOT BSM schema on DTG)
4. coreData_accelset_lat
5. coreData_accelset_long
6. coreData_accelset_vert
7. coreData_accelset_accelYaw (named this way to be consistent with WYDOT BSM schema on DTG)

**Pull request merged**